aptos-move/framework/src/natives/cryptography/algebra/serialization.rs

```diff
@@ -0,0 +1,398 @@
+        (Some(Structure::BLS12381Gt), Some(SerializationFormat::BLS12381Gt)) => {
+            let buf = ark_serialize_internal!(
+                context,
+                args,
+                ark_bls12_381::Fq12,
+                serialize_uncompressed,
+                gas_params.ark_bls12_381_fq12_serialize * NumArgs::one()
+            );
+            Ok(smallvec![Value::vector_u8(buf)])
+        },
+        _ => Err(SafeNativeError::Abort {
+            abort_code: MOVE_ABORT_CODE_NOT_IMPLEMENTED,
+        }),
+    }
+}
+
+/// Macros that implements `deserialize_internal()` using arkworks libraries.
+macro_rules! ark_deserialize_internal {
+    ($context:expr, $bytes:expr, $ark_typ:ty, $ark_deser_func:ident, $gas:expr) => {{
+        $context.charge($gas)?;
+        match <$ark_typ>::$ark_deser_func($bytes) {
+            Ok(element) => {
+                let handle = store_element!($context, element);
+                Ok(smallvec![Value::bool(true), Value::u64(handle as u64)])
+            },
+            Err(ark_serialize::SerializationError::InvalidData)
+            | Err(ark_serialize::SerializationError::UnexpectedFlags) => {
+                Ok(smallvec![Value::bool(false), Value::u64(0)])
+            },
+            _ => {
+                abort_invariant_violated();
+                unreachable!()
+            },
+        }
+    }};
+}
+
+macro_rules! ark_ec_point_deserialize_internal {
+    ($context:expr, $bytes:expr, $typ:ty, $deser_func:ident, $gas:expr) => {{
+        $context.charge($gas)?;
+        match <$typ>::$deser_func($bytes) {
+            Ok(element) => {
+                let element_proj = ark_ec::short_weierstrass::Projective::from(element);
+                let handle = store_element!($context, element_proj);
+                Ok(smallvec![Value::bool(true), Value::u64(handle as u64)])
+            },
+            Err(ark_serialize::SerializationError::InvalidData)
+            | Err(ark_serialize::SerializationError::UnexpectedFlags) => {
+                Ok(smallvec![Value::bool(false), Value::u64(0)])
+            },
+            _ => {
+                abort_invariant_violated();
+                unreachable!()
+            },
+        }
+    }};
+}
+
+pub fn deserialize_internal(
+    gas_params: &GasParameters,
+    context: &mut SafeNativeContext,
+    ty_args: Vec<Type>,
+    mut args: VecDeque<Value>,
+) -> SafeNativeResult<SmallVec<[Value; 1]>> {
+    assert_eq!(2, ty_args.len());
+    let structure_opt = structure_from_ty_arg!(context, &ty_args[0]);
+    let format_opt = format_from_ty_arg!(context, &ty_args[1]);
+    abort_unless_serialization_format_enabled!(context, structure_opt, format_opt);
+    let vector_ref = safely_pop_arg!(args, VectorRef);
+    let bytes_ref = vector_ref.as_bytes_ref();
+    let bytes = bytes_ref.as_slice();
+    match (structure_opt, format_opt) {
+        (Some(Structure::BLS12381Fr), Some(SerializationFormat::BLS12381FrLsb)) => {
+            if bytes.len() != 32 {
```

Comment:
The byte lengths could be constants for readability

Reply:
IMO it is part of the deserialization scheme definition and here we are implementing the definition...
So not a magic number shared across the codebase, and doesn't need to be pulled out as a constant.

I will put some comments around instead.